packages/nimble-components/src/table/specs/data-hierarchy-hld.md

```diff
@@ -59,6 +53,68 @@ interface TableRowExpansionToggleEventDetail {
+    (typeof TableRecordDelayedHierarchyState)[keyof typeof TableRecordDelayedHierarchyState];
+
+public Table() {
+    ...
+    // Sets the hierarchy options for the rows specified by the passed IDs.
+    public async setRecordHierarchyOptions(rowHierarchyOptions: { id: string, options: TableRecordHierarchyOptions }[]): Promise<void>;
+}
+```
+
+The loading state of a row will look as follows:
+
+![Delayed Loading Spinner](./spec-images/LazyLoadingSpinner.gif)
+
+Some notes about the `setRecordHierarchyOptions` API:
+
+-   all options will be cleared when the table's `idFieldName` changes
+-   an option passed to `setRecordHierarchyOptions` with an ID that does not match a record in the data will be ignored
+-   the options passed to `setRecordHierarchyOptions` will override any options previously set to become the complete set of options configured on the table
+-   all options will be ignored if `parentIdFieldName` is not configured on the table
+-   the table will not render delayed hierarchy state (loading or expandable) if the table's `parentIdFieldName` is not configured; however, the options will remain cached within the table if the `parentIdFieldName` becomes `undefined`, and that cached configuration will render in the table if the table's `parentIdFieldName` is changed back to a non-`undefined` value
+-   calling `setData` will clear options associated with IDs that are no longer present in the data
+
+The expected usage of the dynamically loaded hierarchy is as follows:
+
+1. Configure `idFieldName` and `parentIdFieldName` on the table
+1. Call `setData` on the table with the records that are known
+1. Call `setRecordHierarchyOptions` with an option of `{ delayedHierarchyState: TableRecordDelayedHierarchyState.canLoadChildren }` for IDs associated with records that are known to have children that have not been loaded.
+1. Handle the `row-expansion-toggle` event on the table by doing the following for rows that need to dynamically load their data:
+    1. Start loading the data in the background.
+    1. Call `setRecordHierarchyOptions` on the table to update the state of the row whose data is being loaded to `{ delayedHierarchyState: TableRecordDelayedHierarchyState.loadingChildren }`. This call must include the current state of any records not in the default state.
+1. When the dynamically loaded data has finished loading, do the following based on the result of loading data:
+    - If loading the data succeeds:
+        1. Call `setData` on the table with the new data
+        1. Call `setRecordHierarchyOptions` with the current row options for each row that is either still loading or still has data that hasn't been loaded yet.
+    - If loading the data fails but can be tried again:
+        1. Call `setRecordHierarchyOptions` to change the row state back to `{ delayedHierarchyState: TableRecordDelayedHierarchyState.canLoadChildren }`. This call must include the current state of any records not in the default state.
```

Comment:
What happens to the expanded state of that row in this scenario? Does calling setData clear the current expanded states since those rows don't have children so why would they be expanded?

Reply:
I imagined that the code for this would live within the implementation of `setRecordHierarchyOptions`, and if a row is being set to `canLoadChildren` without actually having any children, we'd tell the `ExpansionManager` to ensure the row is collapsed. In this scenario, there aren't any calls being made to `setData` (other than setting the initial table data), so I don't think that logic would play into this behavior.

Reply:
Although, I think you're right that there is another case to handle here, which is a `setData` call removes all children of an expanded row that as `canLoadChildren` set on it. In that case, we should also make sure the row ends up collapsed again.

In summary, a row with no children in the data and `canLoadChildren` set on it needs to be collapsed, no matter what code path gets into that state.